### PR TITLE
Comments: Display Counts

### PR DESCRIPTION
**Pull request by gwwar**
This PR displays comment counts, and shows a localized thousands compact variant if available. If we don't provide one, or such a shorthand does not exist for the locale it falls back to our usual number localization. 

screen shot 2018-01-27 at 5 12 44 pm

screen shot 2018-01-27 at 5 20 04 pm

screen shot 2018-01-27 at 5 20 34 pm

screen shot 2018-01-30 at 3 57 22 pm

### Testing Instructions
- Navigate to /comments
- Select a site with < 1000 comments
- Perform moderation actions, comment counts should update appropriately
- Select a site with > 1000 comments
- We should see expected shortening of counts as shown in images.

### CODE REVIEW

File: client/my-sites/comments/comment-navigation/index.jsx [hunk 1]
```diff
@@ -51,6 +51,29 @@ const bulkActions = {
...
+ *
+ * @example
+ * countAbbreviation( 999 ) returns 999
+ * countAbbreviation( 1234 ) returns 1.2K
+ * countAbbreviation( 27892 ) returns 28K
+ *
+ * @param {Number} count - raw count
+ * @returns {String} - count string
+ */
+const countAbbreviation = function( count ) {
```
Comment (rodrigoi) [inline client/my-sites/comments/comment-navigation/index.jsx]: we need to support localization for this.

screen shot 2018-01-28 at 02 24 00
screen shot 2018-01-28 at 02 27 20
screen shot 2018-01-28 at 02 28 59

For example, in Spanish the correct abbreviation is `1,7K`. Maybe this will be a better job for `numberFormat`, by adding a `thousands_abbreviation_translation_key` to our translations, with a `compact` option.

Comment (Copons) [reply client/my-sites/comments/comment-navigation/index.jsx]: This is a great catch.
Italy uses the comma as well, but I'm just so used at seeing both formats that I had a hard time realizing it was incorrect.

Although I've tried to wire up `numberFormat` quickly and dirtily and it looks a bit off:

```es6
// Show 2 sig figs
if ( count < 10000 ) {
	return `${ this.props.numberFormat( count, { decimals: 1, thousandsSep: '' } ) }K`;
}
```

It doesn't really acknowledge that my locale is `it`, and keeps using the period as decimal separator instead of the comma.
Also it disregards `thousandsSep: ''` as falsy, showing the decimal separator anyway (this time it's a period, which is correct for my locale).

Comment (rodrigoi) [reply client/my-sites/comments/comment-navigation/index.jsx]: I was thinking of something more like:
```jsx
<Count count={ count } compact />
```
and then
```js
const { compact } = this.props;
{ numberFormat( count, { compact } ) }
```
The problem is not just the comma. Some countries use `M` and `Mil` instead of `K`. Localization is hard 😛 .

Comment (gwwar) [reply client/my-sites/comments/comment-navigation/index.jsx]: The thousands separator should be easy enough to work around, but are we supporting other variants for this round vs plain metric? @drw158 @megsfulton I know India for example uses completely different counting units.

Comment (Copons) [reply client/my-sites/comments/comment-navigation/index.jsx]: Absolutely!

My point was that, currently, `i18n.numberFormat` has got issues messing with what we need.
It apparently has no way to hide the thousands separator, and it doesn't use the correct decimal separator according to the locale.

Anyways, I went to see how Twitter handles this for some random languages, and...

| English | Italian | Spanish |
| --- | --- | --- |
| screen shot 2018-01-29 at 16 46 47 | screen shot 2018-01-29 at 16 47 32 | screen shot 2018-01-29 at 16 47 51 |



> Localization is hard

👍

Comment (Copons) [reply client/my-sites/comments/comment-navigation/index.jsx]: I'd say let's ship this with the the period and the K for now, and then work with the i18n folks to add a compact option to `numberFormat`.

Comment (gwwar) [reply client/my-sites/comments/comment-navigation/index.jsx]: Hmm, we could also just save some design time and do what twitter does... I'll try giving it another shot after meetings today, and we can loop in the globalization team to check it

Comment (Copons) [reply client/my-sites/comments/comment-navigation/index.jsx]: It's just that I've tried 3 different languages and I've got 3 different formats.
I can't even imagine how many possibilities we could find when starting to support all languages supported by Calypso.

Like off the top of my head, a "compact" number in any given locale can have:
- `.` or `,` as decimal separator
- `,` or `.` or _nothing_ as thousands separator
- `K`, `M`, `mil`, or any other Unicode character, hopefully emoji excluded, as "thousands" abbreviation
- or... be "expanded" instead (as it happens in Italian).

That's not something that we can solve with some design time, isn't it?

Comment (gwwar) [reply client/my-sites/comments/comment-navigation/index.jsx]: I was mostly thinking we could match output for our top supported languages. I had to do something similar for [currency formatting](https://github.com/Automattic/wp-calypso/tree/master/client/lib/format-currency). Though we should probably add this option in the i18n module.

**Comment by gwwar** [comment]
I'll drop bbf0b8e for now since I'll likely need to update i18n-calypso first.

**Comment by gwwar** [comment]
Hmm just showing bare counts forces this into a dropdown if we have a ton of comments.
screen shot 2018-01-29 at 5 19 31 pm

I'll try to prep some i18n updates.

**Comment by megsfulton** [comment]
Showing the count is a nice improvement 👏 

I'll defer to @drw158 but I think 'K' is a safe fallback.

**Comment by gwwar** [comment]
Before I get too far in filling this out, @rodrigoi @Copons how does the approach in edaaf47 look?

File: client/lib/format-number-compact/index.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Formats a number to a short format given a language code
+ * @param   {Number}     number              number to format
+ * @param   {String}     code                language code e.g. 'es'
+ * @returns {?String}                        A formatted string.
+ */
+export default function formatNumberCompact( number, code ) {
+	//use numberFormat directly from i18n in this case!
+	if ( isNaN( number ) || ! THOUSANDS[ code ] ) {
+		return null;
```
Comment (Copons) [inline client/lib/format-number-compact/index.js]: Not exactly sure how this would work.
Since we have `numberFormat` available, why not directly returning `numberFormat( number )` (which is what I'd expect from the inline comment above, btw)?

**Comment by Copons** [comment]
> Hmm just showing bare counts forces this into a dropdown if we have a ton of comments.

Not much we can do there. 🙁 
The navbar automagically compares the tabs width against the navbar available space, and collapse the tabs into a dropdown if needed.
We have many tabs, and two buttons on the right (which can be quite long, depending on the language, e.g. in Italian "bulk edit" is an ~twice longer "modifica di massa").

Anyways, imho the pros of showing the counts >>>> the cons of collapsing into a dropdown more often.

> how does the approach in 703b59a look?

Fine by me!
I've left a comment, but overall it looks good. I guess we could internally "crowdsource" the formats for the top languages in another PR.

One question: what about languages like Italian (which, admittedly, could be just Italian 😄) that don't use the compact version?
What about something like this:
```es6
it: {
	symbol: '', //empty string
	grouping: '.',
	decimal: ',',
}
```
and handling the empty `symbol` as a "not compact", while still using the grouping and decimal values?

**Comment by gwwar** [comment]
> One question: what about languages like Italian (which, admittedly, could be just Italian 😄) that don't use the compact version?

@Copons I tried that yesterday, and I found that its redundant to mark it in this config file, when we do want to use numberFormat as the fallback. It does work though :)

I was thinking in count we could have a generic string override, until I port this over to i18n-calypso and add a compact option to numberFormat.

> Since we have numberFormat available

The numberFormat that count uses has [additional state](https://github.com/Automattic/i18n-calypso/blob/9ba44fa6482fc15b9c7782d5fc957608eb3db521/lib/index.js#L196) available from the i18n instance.

**Comment by gwwar** [comment]
I'll add a few more locales, but wanted to ping global for a look.

**Comment by akirk** [comment]
It looks good and sounds useful. Getting more locale support can be a bit painful but you can use CLDR as a source for locale abbreviations, for example (`numbers['decimalFormats-numberSystem-latn'].short.decimalFormat`)
- https://github.com/unicode-cldr/cldr-numbers-full/blob/master/main/de/numbers.json
- https://github.com/unicode-cldr/cldr-numbers-full/blob/master/main/it/numbers.json

It's even correct for what @Copons said :)

Given the number of locales it might make sense to write a script to extract the values from CLDR and transform them into your format.

**Comment by gwwar** [comment]
Thanks @akirk! I'll see what I can do. So looking at `it` it should be `mille`? But I suppose twitter drops it because it's longer than the original?

**Comment by akirk** [comment]
Hm, not sure we're looking at the same data. Here's English ([en/numbers.json](https://github.com/unicode-cldr/cldr-numbers-full/blob/master/main/en/numbers.json)):
```
              "1000-count-other": "0K",
              "10000-count-other": "00K",
              "100000-count-other": "000K",
              "1000000-count-other": "0M",
```
Compare with Italian:
```
              "1000-count-other": "0",
              "10000-count-other": "0",
              "100000-count-other": "0",
              "1000000-count-other": "0 Mln",
```
Compare with German:
```
              "1000-count-other": "0 Tsd'.'",
              "10000-count-other": "0 Tsd'.'",
              "100000-count-other": "0 Tsd'.'",
              "1000000-count-other": "0 Mio'.'",
```
I read this, as @Copons said, no abbreviation for thousands in Italian.

**Comment by gwwar** [comment]
@akirk Yeah I was looking at the `long` format for `it`

**Comment by akirk** [comment]
Maybe it could be helpful to take a look at the number formatter of globalize (which uses CDLR data), see https://github.com/globalizejs/globalize#number-module

File: client/lib/format-number-compact/thousands.js [hunk 3]
```diff
@@ -10,4 +10,14 @@ export const THOUSANDS = {
 		grouping: '.',
 		decimal: ',',
 	},
+	'pt-br': {
+		symbol: ' mil',
+		grouping: '.',
+		decimal: ',',
+	},
+	de: {
+		symbol: ' Tsd',
```
Comment (akirk) [inline client/lib/format-number-compact/thousands.js]: The symbol should actually be `Tsd.`

Comment (gwwar) [reply client/lib/format-number-compact/thousands.js]: updated in f0ec9a0

**Comment by gwwar** [comment]
@akirk Any catches are welcome :) I'm not super sure if I got down `ar` correctly, as I was fighting with my IDE.

I've added all of our popular locales, any missing from the file should default to what we normally use in `numberFormat`

**Comment by Copons** [comment]
> So looking at `it` it should be `mille`?

"mille" is literally "one thousand" (singular), and "N mila" is used like "N thousands" (plural).

1000 = mille
2000 = duemila (2-mila)
3000 = tremila (3-mila)

Though, it's not an abbreviation, but the actual full name of those numbers.
And in fact, [here](https://github.com/unicode-cldr/cldr-numbers-full/blob/master/main/it/numbers.json) it uses them for the `long` formats.
The `short` formats don't use any symbol until the million, and then:
- `Mln`, "milione" (million)
- `Mld`, "miliardo" (billion)
- `Bln`, "bilione" (which sounds like billion, but it's indeed 1000 billions - which is not a trillion isn't it? 😄)

So, in other words this is all very scary and I second the proposal of relying on external data.

**Comment by kwight** [comment]
Bless your hearts for fighting the good fight <3

**Comment by gwwar** [comment]
Added https://github.com/Automattic/wp-calypso/issues/22010 too so showing unabbreviated counts won't be as painful

**Comment by gwwar** [comment]
@akirk @ramonjd any recommendations for next steps here? Is this change set okay, or would folks like changes?

**Comment by ramonjd** [comment]
Ran and tested it. I think it's a good start and sufficiently germane to create one change set, though I'll defer to @akirk

Could be overcomplicating things, but maybe in a future iteration we could fallback to the root language abbreviation for variants (assuming they're not wildly divergent)? 

**French**
screen shot 2018-02-01 at 2 31 45 pm

**French Canadian**
screen shot 2018-02-01 at 2 31 20 pm

**Comment by gwwar** [comment]
> but maybe in a future iteration we could fallback to the root language abbreviation

I think that makes sense @ramonjd ! I'll leave that for another PR though

Thanks for the reviews @akirk, @ramonjd, @rodrigoi, @Copons !
